src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+  int n_nodes = _body_nodes.length();
+  Node_List* s2v_map = new Node_List(_arena, n_nodes);
+
+  // Replace each node with valid element basic type set
+  for (int idx = 0; idx < n_nodes; idx++) {
+    Node* snode = _body_nodes.at(idx);
+    if (has_valid_elem_bt(snode)) {
+      Node* vnode;
+      int opc = snode->Opcode();
+      BasicType bt = elem_bt(snode);
+      int vlen = Matcher::max_vector_size(bt);
+      if (snode->is_Mem()) {
+        Node* ctrl = snode->in(MemNode::Control);
+        Node* mem = snode->in(MemNode::Memory);
+        Node* addr = snode->in(MemNode::Address);
+        const TypePtr* at = snode->as_Mem()->adr_type();
+        const TypeVect* vt = TypeVect::make(Type::get_const_basic_type(bt), vlen);
+        if (snode->is_Load()) {
+          vnode = new LoadVectorMaskedNode(ctrl, mem, addr, at, vt, mask);
+        } else {
+          assert(snode->is_Store(), "Unexpected memory op");
+          Node* val = get_vector_input(snode, MemNode::ValueIn);
+          vnode = new StoreVectorMaskedNode(ctrl, mem, addr, val, at, mask);
+        }
+      } else if (VectorNode::is_convert_opcode(opc)) {
+        Node* in = get_vector_input(snode, 1);
+        int vopc = VectorCastNode::opcode(opc, in->bottom_type()->is_vect()->element_basic_type());
+        vnode = VectorCastNode::make(vopc, in, bt, vlen);
+      } else {
+        uint start, end;
+        VectorNode::vector_operands(snode, &start, &end);
+        assert(start == 1, "Start should be 1 for all currently supported ops");
+        // The 1st operand is always there
+        Node* in1 = get_vector_input(snode, 1);
+        // The 2nd operand is optional and may be vector shift count
+        Node* in2 = nullptr;
+        if (end > 2 || VectorNode::is_shift(snode) || VectorNode::is_roundopD(snode)) {
+          in2 = get_vector_input(snode, 2);
+        }
+        // The 3rd operand is optional
+        if (end > 3) {
+          Node* in3 = get_vector_input(snode, 3);
+          vnode = VectorNode::make(opc, in1, in2, in3, vlen, bt);
+        } else {
+          vnode = VectorNode::make(opc, in1, in2, vlen, bt);
+        }
+      }
+      VectorNode::trace_new_vector(vnode, "VectorMasked");
+      _phase->set_ctrl(vnode, _phase->get_ctrl(snode));
+      _igvn->replace_node(snode, _igvn->register_new_node_with_optimizer(vnode, snode));
+      s2v_map->map(rpo_idx(snode), vnode);
+    }
+  }
+
+#ifndef PRODUCT
+  if (TraceMaskedLoop) {
+    tty->print_cr("Node scalar to vector replacements");
+    for (int idx = 0; idx < _body_nodes.length(); idx++) {
+      Node* snode = _body_nodes.at(idx);
+      if (has_valid_elem_bt(snode)) {
+        Node* vnode = s2v_map->at(rpo_idx(snode));
+        tty->print(" Scalar:\t");
+        snode->dump();
+        tty->print("  Vector:\t");
+        vnode->dump();
+      }
+    }
+    tty->cr();
+  }
+#endif
+
+  return s2v_map;
+}
+
+// Duplicate vectorized operations with given vector element size
```

Comment:
Got to here today. There should probably be some comment higher up that you first replace scalars with one vector each, and then duplicate them for the larger types that need multiple vectors.

I'm also concerned that there may be some platforms where the max vector width in bytes is not the same for all types. But maybe all platforms that support masked register ops also all have the same vector width in bytes for all types?

Reply:
Assume we only allow `32` bit registers for `int`, but `64` bits for doubles. Now you'd be assuming that there need to be double as many `double` vectors as `int` vectors. But actually, they need the same amount of vectors, because vectors of both sizes fit exactly `8` elements.

Reply:
More comments are added. I can only say this way is good on current AArch64. As we don't have enough knowledge of other architectures, we may need some help if we need to change this.